File: batch/ats-scanner.mjs
import fs from 'fs';
import path from 'path';
import https from 'https';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));

/**
 * Resolve vault path from env var or config.
 * Priority: CAREER_OPS_VAULT_PATH env > obsidian-bridge.md default
 */
function resolveVaultPath() {
  if (process.env.CAREER_OPS_VAULT_PATH) return process.env.CAREER_OPS_VAULT_PATH;
  const home = process.env.HOME || process.env.USERPROFILE || '';
  return path.join(home, 'Library/Mobile Documents/iCloud~md~obsidian/Documents/Life_OS/10_Projects/02_Job_Hunt_2026');
}

const VAULT_PATH = resolveVaultPath();
const JD_POOL = path.join(VAULT_PATH, '10_JD_Pool');

// Known ATS mappings for major targets (Lever and Greenhouse)
const ATS_TARGETS = [
  // --- Original targets ---
  { name: 'Stripe', type: 'greenhouse', slug: 'stripe' },
  { name: 'Figma', type: 'greenhouse', slug: 'figma' },
  { name: 'Canva', type: 'lever', slug: 'canva' },
  { name: 'Discord', type: 'greenhouse', slug: 'discord' },
  { name: 'Airbnb', type: 'greenhouse', slug: 'airbnb' },
  { name: 'Pinterest', type: 'greenhouse', slug: 'pinterest' },
  { name: 'Reddit', type: 'greenhouse', slug: 'reddit' },
  { name: 'Riot Games', type: 'greenhouse', slug: 'riotgames' },
  { name: 'Epic Games', type: 'greenhouse', slug: 'epicgames' },
  { name: 'GitLab', type: 'greenhouse', slug: 'gitlab' },
  { name: 'Twitch', type: 'greenhouse', slug: 'twitch' },
  { name: 'The Trade Desk', type: 'greenhouse', slug: 'thetradedesk' },
  { name: 'Criteo', type: 'greenhouse', slug: 'criteo' },
  { name: 'Agoda', type: 'greenhouse', slug: 'agabordaresort' },
  { name: 'Notion', type: 'lever', slug: 'notion' },
  { name: 'Roblox', type: 'greenhouse', slug: 'roblox' },
  // --- APAC Expansion: Fintech/Payments ---
  { name: 'Airwallex', type: 'greenhouse', slug: 'airwallex' },
  { name: 'Wise', type: 'greenhouse', slug: 'transferwise' },
  { name: 'Adyen', type: 'greenhouse', slug: 'adyen' },
  { name: 'Circle', type: 'greenhouse', slug: 'circle' },
  { name: 'Revolut', type: 'lever', slug: 'revolut' },
  // --- APAC Expansion: Developer Tools / SaaS ---
  { name: 'Datadog', type: 'greenhouse', slug: 'datadog' },
  { name: 'Cloudflare', type: 'greenhouse', slug: 'cloudflare' },
  { name: 'Elastic', type: 'greenhouse', slug: 'elastic' },
  { name: 'MongoDB', type: 'greenhouse', slug: 'mongodb' },
  { name: 'Confluent', type: 'greenhouse', slug: 'confluent' },
  { name: 'HashiCorp', type: 'greenhouse', slug: 'hashicorp' },
  { name: 'Grafana Labs', type: 'greenhouse', slug: 'grafanalabs' },
  { name: 'Sentry', type: 'greenhouse', slug: 'sentry' },
  { name: 'PostHog', type: 'greenhouse', slug: 'posthog' },
  { name: 'Linear', type: 'greenhouse', slug: 'linear' },
  // --- APAC Expansion: Crypto/Web3 ---
  { name: 'Kraken', type: 'greenhouse', slug: 'kraken' },
  { name: 'Chainlink Labs', type: 'greenhouse', slug: 'chainlinklabs' },
  { name: 'Consensys', type: 'greenhouse', slug: 'consensys' },
  // --- APAC Expansion: Gaming ---
  { name: 'Ubisoft', type: 'greenhouse', slug: 'ubisoft' },
  { name: 'Electronic Arts', type: 'greenhouse', slug: 'electronicarts' },
  // --- APAC Expansion: Travel/Marketplace ---
  { name: 'Klook', type: 'greenhouse', slug: 'klook' },
  { name: 'PropertyGuru', type: 'greenhouse', slug: 'propertyguru' },
  // --- APAC Expansion: Enterprise ---
  { name: 'ServiceNow', type: 'greenhouse', slug: 'servicenow' },
  { name: 'Twilio', type: 'greenhouse', slug: 'twilio' },
];

const TARGET_LOCATIONS = ['singapore', 'beijing', 'shanghai', 'hong kong', 'shenzhen', 'hangzhou', 'guangzhou', 'chengdu', 'remote', 'apac', 'asia'];

function stripHtml(html) {
  if (!html) return '';
  return html.replace(/<[^>]*>?/gm, '')
             .replace(/&nbsp;/g, ' ')
             .replace(/&amp;/g, '&')
             .replace(/\n\s*\n/g, '\n\n')
             .trim();
}

const fetchUrl = (url) => {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'Mozilla/5.0' } }, (res) => {
      if (res.statusCode !== 200) {
        return reject(`Failed to fetch ${url}: ${res.statusCode}`);
      }
      let rawData = '';
      res.on('data', (chunk) => { rawData += chunk; });
      res.on('end', () => {
        try { resolve(JSON.parse(rawData)); } catch(e) { reject(e); }
      });
    }).on('error', reject);
  });
};

async function processGreenhouse(company) {
  const url = `https://boards-api.greenhouse.io/v1/boards/${company.slug}/jobs?content=true`;
  console.log(`[${company.name}] Scanning Greenhouse API...`);
  try {
    const data = await fetchUrl(url);
    if (!data.jobs) return;
    
    for (const job of data.jobs) {
      const title = job.title.toLowerCase();
      const loc = job.location.name.toLowerCase();
      
      // Filter: Must be a Product Manager role in Target Locations
      if ((title.includes('product manager') || title.includes('product leader')) && !title.includes('engineering')) {
        const matchLoc = TARGET_LOCATIONS.some(tl => loc.includes(tl));
        if (matchLoc || loc.includes('china') || loc.includes('apac')) {
          saveToVault(company.name, job.title, job.location.name, job.absolute_url, job.id, stripHtml(job.content));
        }
      }
    }
  } catch (err) {
    console.error(`[${company.name}] API Error: ${err}`);
  }
}

async function processLever(company) {
  const url = `https://api.lever.co/v0/postings/${company.slug}?mode=json`;
  console.log(`[${company.name}] Scanning Lever API...`);
  try {
    const data = await fetchUrl(url);
    for (const job of data) {
      const title = job.text.toLowerCase();
      const loc = job.categories?.location?.toLowerCase() || '';
      const content = job.description || job.descriptionPlain || '';
      
      if ((title.includes('product manager') || title.includes('product leader'))) {
        const matchLoc = TARGET_LOCATIONS.some(tl => loc.includes(tl));
        if (matchLoc || loc.includes('china') || loc.includes('apac')) {
          saveToVault(company.name, job.text, job.categories.location, job.hostedUrl, job.id, stripHtml(content));
        }
      }
    }
  } catch (err) {
    console.error(`[${company.name}] API Error: ${err}`);
  }
}

function saveToVault(company, title, location, url, id, content) {
  const safeTitle = title.replace(/\//g, '-').replace(/:/g, '');
  const fileName = `${safeTitle} | ${company} | ATS.md`;
  const filePath = path.join(JD_POOL, fileName);

  const markdown = `---
title: "${safeTitle} | ${company} | ATS"
source: "${url}"
created: ${new Date().toISOString().split('T')[0]}
tags:
  - jobs
  - ats-scanned
status: new
Tier: ""
score: ""
company: "${company}"
location: "${location}"
remote: "${location.toLowerCase().includes('remote') ? 'remote' : 'hybrid'}"
archetype: ""
pdf_generated: false
description: "Automatically extracted via ATS API proxy."
_ats_id: "${id}"
---

## ${title}

**Location:** ${location}
**Company:** ${company}
**ATS JD ID:** \`${id}\`

### Job Description
${content}
`;

  fs.writeFileSync(filePath, markdown);
  console.log(`  -> Saved: ${fileName}`);
}

async function main() {
  console.log(`Starting ATS Scanner... Mapping ${ATS_TARGETS.length} Companies`);
  for (const company of ATS_TARGETS) {
    if (company.type === 'greenhouse') {
      await processGreenhouse(company);
    } else if (company.type === 'lever') {
      await processLever(company);
    }
  }
  console.log('ATS Scanning Complete.');
}

main();

File: batch/ats-scanner.py
import urllib.request
import json
import os
import re
from datetime import datetime

VAULT_PATH = '/Users/boss/Library/Mobile Documents/iCloud~md~obsidian/Documents/Life_OS/10_Projects/02_Job_Hunt_2026'
JD_POOL = os.path.join(VAULT_PATH, '10_JD_Pool')

ATS_TARGETS = [
    { 'name': 'Stripe', 'type': 'greenhouse', 'slug': 'stripe' },
    { 'name': 'Figma', 'type': 'greenhouse', 'slug': 'figma' },
    { 'name': 'Canva', 'type': 'greenhouse', 'slug': 'canva' }, # Try GH first
    { 'name': 'Discord', 'type': 'greenhouse', 'slug': 'discord' },
    { 'name': 'Airbnb', 'type': 'greenhouse', 'slug': 'airbnb' },
    { 'name': 'Pinterest', 'type': 'greenhouse', 'slug': 'pinterest' },
    { 'name': 'Reddit', 'type': 'greenhouse', 'slug': 'reddit' },
    { 'name': 'Riot Games', 'type': 'greenhouse', 'slug': 'riotgames' },
    { 'name': 'Epic Games', 'type': 'greenhouse', 'slug': 'epicgames' },
    { 'name': 'GitLab', 'type': 'greenhouse', 'slug': 'gitlab' },
    { 'name': 'Twitch', 'type': 'greenhouse', 'slug': 'twitch' },
    { 'name': 'The Trade Desk', 'type': 'greenhouse', 'slug': 'thetradedesk' },
    { 'name': 'Agoda', 'type': 'greenhouse', 'slug': 'agoda' },
    { 'name': 'Roblox', 'type': 'greenhouse', 'slug': 'roblox' },
    { 'name': 'Notion', 'type': 'lever', 'slug': 'notion' },
    { 'name': 'Shopee', 'type': 'lever', 'slug': 'shopee' }
]

TARGET_LOCATIONS = ['singapore', 'beijing', 'shanghai', 'hong kong', 'remote']

def strip_html(text):
    if not text: return ''
    text = re.sub(r'<[^>]*>?', '', text)
    text = text.replace('&nbsp;', ' ').replace('&amp;', '&')
    text = re.sub(r'\n\s*\n', '\n\n', text)
    return text.strip()

def fetch_json(url):
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    try:
        with urllib.request.urlopen(req) as response:
            return json.loads(response.read().decode())
    except Exception as e:
        return None

def save_to_vault(company, title, location, url, job_id, content):
    safe_title = title.replace('/', '-').replace(':', '')
    file_name = f"{safe_title} | {company} | ATS.md"
    file_path = os.path.join(JD_POOL, file_name)
    
    date_str = datetime.now().strftime('%Y-%m-%d')
    remote_flag = 'remote' if 'remote' in location.lower() else 'hybrid'
    
    markdown = f"""---
title: "{safe_title} | {company} | ATS"
source: "{url}"
created: {date_str}
tags:
  - jobs
  - ats-scanned
status: new
Tier: ""
score: ""
company: "{company}"
location: "{location}"
remote: "{remote_flag}"
archetype: ""
pdf_generated: false
description: "Automatically extracted via ATS API proxy."
_ats_id: "{job_id}"
---

## {title}

**Location:** {location}
**Company:** {company}
**ATS JD ID:** `{job_id}`

### Job Description
{content}
"""
    try:
        with open(file_path, 'w') as f:
            f.write(markdown)
        print(f"  -> Saved: {file_name}")
    except Exception as e:
        print(f"Failed to save {file_name}: {e}")

def process_greenhouse(company):
    url = f"https://boards-api.greenhouse.io/v1/boards/{company['slug']}/jobs?content=true"
    print(f"[{company['name']}] Scanning Greenhouse API...")
    data = fetch_json(url)
    if not data or 'jobs' not in data: return
    
    for job in data['jobs']:
        title = job.get('title', '').lower()
        loc = job.get('location', {}).get('name', '').lower()
        
        if ('product manager' in title or 'product leader' in title) and 'engineering' not in title:
            if any(tl in loc for tl in TARGET_LOCATIONS) or 'china' in loc or 'apac' in loc:
                content = strip_html(job.get('content', ''))
                save_to_vault(company['name'], job['title'], job['location']['name'], job['absolute_url'], job.get('id', ''), content)

def process_lever(company):
    url = f"https://api.lever.co/v0/postings/{company['slug']}?mode=json"
    print(f"[{company['name']}] Scanning Lever API...")
    data = fetch_json(url)
    if not data: return
    
    for job in data:
        title = job.get('text', '').lower()
        loc = job.get('categories', {}).get('location', '').lower()
        
        if ('product manager' in title or 'product leader' in title):
            if any(tl in loc for tl in TARGET_LOCATIONS) or 'china' in loc or 'apac' in loc:
                content = strip_html(job.get('description', '') or job.get('descriptionPlain', ''))
                save_to_vault(company['name'], job['text'], loc, job.get('hostedUrl', ''), job.get('id', ''), content)

def main():
    print(f"Starting ATS Scanner... Mapping {len(ATS_TARGETS)} Companies")
    os.makedirs(JD_POOL, exist_ok=True)
    
    for company in ATS_TARGETS:
        if company['type'] == 'greenhouse':
            process_greenhouse(company)
        elif company['type'] == 'lever':
            process_lever(company)
    print("ATS Scanning Complete.")

if __name__ == "__main__":
    main()
